stripe/src/main/java/com/stripe/android/paymentsheet/PaymentOptionResult.kt

```diff
@@ -17,18 +17,23 @@ internal sealed class PaymentOptionResult(
         return bundleOf(EXTRA_RESULT to this)
     }
 
-    @Parcelize
-    data class Succeeded(
-        val paymentSelection: PaymentSelection,
-        val newSavedPaymentMethod: PaymentMethod? = null,
-    ) : PaymentOptionResult(Activity.RESULT_OK) {
-        init {
-            val isSavedNewCard =
-                (paymentSelection as? PaymentSelection.New.Card)?.shouldSavePaymentMethod == true
-            if (isSavedNewCard) {
-                requireNotNull(newSavedPaymentMethod)
-            }
-        }
+    internal sealed class Succeeded : PaymentOptionResult(Activity.RESULT_OK) {
+        abstract val paymentSelection: PaymentSelection
+
+        /**
+         * This could either be a saved payment method or a new payment method that is not
+         * saved.
+         */
+        @Parcelize
+        data class NotNewlySaved(
```

Comment:
What do you think of splitting these out to two separate types, `Unsaved` and `Existing`?

Reply:
There are basically three cases here:
1. The card is new and will not be saved, in which case we just need to return the selection
2. The card is already saved, and we just need to return the selection
3. The card is new and was saved, and now we need to save the new payment method in the flow controller view model.

So since a new unsaved card doesn't have a payment method it would have to be `Existing`, which doesn't work.

The other challenge is that in PaymentOptionsViewModel.getPaymentResult needs to return one of these results.  The method is called in the case of a cancellation in order to populate the last selection in the launcher.  The behavior is that whatever the last full card shown is in view, is the last reported selection in the launcher, thus it must return some form of Succeeded if there is any selection.

What about something like WithNewMethod and WithNoNewMethod?  Kinda verbose?

Reply:
I think my comment wasn't clear. I was suggesting splitting `NotNewlySaved` out to `Unsaved` and `Existing`. `NewlySaved` would be a separate type. So ultimately `Succeeded` would have 3 subclasses.

Reply:
sure!